keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1636 @@
+    // the ResourceClaim.
+    Spec ResourceClaimSpec
+    // Status describes whether the resource is available and with
+    // which attributes.
+    Status ResourceClaimStatus
+}
+
+type ResourceClaimSpec struct {
+    // ResourceClassName references the driver and additional
+    // parameters via the name of a ResourceClass that was
+    // created as part of the driver deployment.
+    //
+    // The apiserver does not check that the referenced class
+    // exists, but a driver-specific admission webhook
+    // may require that and is allowed to reject claims where
+    // the class is missing.
+    ResourceClassName string
+
+    // Parameters holds arbitrary values that will be available to the driver
+    // when allocating a resource for the claim.
+    Parameters runtime.RawExtension
+
+    // Allocation can start immediately or when a Pod wants to use
+    // the resource. Waiting for a Pod is the default.
+    AllocationMode AllocationMode
+}
+
+// AllocationMode describes whether a ResourceClaim gets allocated
+// immediately when it gets created (AllocationModeImmediate)
+// or whether allocation is delayed until it is needed for
+// a Pod (AllocationModeDelayed). Other modes might get
+// added in the future.
+type AllocationMode string
+
+const (
+     // When a ResourceClass has AllocationModeImmediate,
+     // allocation of ResourceClaims with that ResourceClass
+     // starts as soon as the ResourceClaim gets created.
+     // This is done without considering the needs of
+     // Pods that will use the ResourceClaim because those
+     // Pods are not known yet.
+     AllocationModeImmediate AllocationMode = “Immediate”
+
+     // When a ResourceClass has AllocationModeDelayed,
+     // allocation of ResourceClaims with that ResourceClass
+     // is delayed until a Pod gets scheduled that needs the
+     // ResourceClaim. The scheduler will consider all
+     // resource requirements of that Pod and trigger
+     // allocation for a node that fits the Pod.
+     AllocationModeDelayed AllocationMode = “Delayed”
+}
+
+type ResourceClaimStatus struct {
+   // Explains what the current status of the claim is and
+   // determines which component needs to do something.
+   Phase ResourceClaimPhase
+
+   // A copy of the driver name from the ResourceClass at
+   // the time when allocation started. Drivers can
+   // filter claims by this field. It's also necessary to
+   // support deallocation when the class gets deleted before
+   // a claim.
+   DriverName string
+
+   // When allocation is delayed, the scheduler must set
+   // the node for which it wants the resource to be allocated
+   // before the driver proceeds with allocation.
+   //
+   // For immediate allocation, the scheduler will not set
+   // this field. The resource driver controller may
+   // set it to trigger allocation on a specific node if the
+   // resources are local to nodes.
+   SelectedNode string
```

Comment:
Does it make sense to abstract some of these fields into sub-structs? I feel as if we need better grouping for the fields that should be observed by the scheduler (vs. written by the scheduler) and the fields that should be observed by the resource driver (vs. written by the resource driver) to make it clearer what their intentions are.

Reply:
That might make sense.

For the initial KEP approval I'd like to keep it as-is, though: we just need enough information about the API to convince ourselves that this can work. The final API is the one we agree upon during API review which (to the best of my knowledge) is not required for a KEP - I've asked for clarification about that today in https://kubernetes.slack.com/archives/CJUQN3E4T/p1643358693199309 because @ahg-g also brought up API review.

Reply:
i agree with @klueska feedback.  grouping items that we anticipate having selective write-authority by actor is helpful.